src/Manager.hs

```diff
@@ -110,9 +140,21 @@ writeMem (Page { frId, memType }) off mem = do
       f <- getFrame @'Swap (Fid $ unFid frId)
       writeFrame f off mem
 
+
 -- Read memory from the page
 readMem :: ManagerSig sig m => Page a -> Offset Word8 -> CountOf Word8 -> m ([Word8])
-readMem p off count = undefined
+readMem (Page {frId, memType}) off count = do
+  case memType of
+    Ram -> do
+      p <- getFrame @'Ram (Fid $ unFid frId)
+      mem <- readFrame p off count
+      return mem
+    Swap -> do
+      p <- getFrame @'Swap (Fid $ unFid frId)
+      mem <- readFrame p off count
+      return mem
```

Comment:
`readFrame ` has `m ([Word8])` return type, you can leave it as last line in the function, without explicitly assigning it to `mem` and returning `mem`